README.md
# glmnetUtils

![R-CMD-check](https://github.com/hongooi73/glmnetUtils/workflows/R-CMD-check/badge.svg)

Some quality-of-life functions to streamline the process of fitting elastic net models with `glmnet`, specifically:

* `glmnet.formula` provides a formula/data frame interface to `glmnet`.
* `cv.glmnet.formula` does a similar thing for `cv.glmnet`.
* Methods for `predict` and `coef` for both the above.
* A function `cva.glmnet` to choose both the alpha and lambda parameters via cross-validation, following the approach described in the help page for `cv.glmnet`. Optionally does the cross-validation in parallel.
* Methods for `plot`, `predict` and `coef` for the above.

You can install the development version from Github using `devtools::install_github`.

    install.packages("devtools")
    library(devtools)
    install_github("hongooi73/glmnetUtils")
    library(glmnetUtils)


NEWS.md
## glmnetUtils 1.1.9

- Remove vignette dependency on an external download.

## glmnetUtils 1.1.8

- Skip some tests on 32-bit Solaris R-patched due to numerical convergence issues.

## glmnetUtils 1.1.7

- Add some plotting options for `plot.cva.glmnet`: `log.x` controls whether to plot the X-axis (lambda) on the log scale, and the legend can be omitted by setting either `legend.x` or `legend.y` to `NULL`.
- Compatibility fixes for glmnet 4.1-1.

## glmnetUtils 1.1.6

- Update maintainer email address.

## glmnetUtils 1.1.5

- Fix handling of non-factor categorical predictors (from R 4.0, data frames will not have character columns converted to factors by default). The practical impact of this should be minor.

## glmnetUtils 1.1.4

- Fix printout of `glmnet.formula` object.

## glmnetUtils 1.1.3

- Support relaxed (non-regularised) fits in `glmnet.formula` and `cv.glmnet.formula` (requires glmnet 3.0 or later).
- Add a legend when plotting a `cva.glmnet` object.

## glmnetUtils 1.1.2

- Fixes a bug in the assignment of observations to crossvalidation folds in `cva.glmnet`. The impact is most serious for small datasets, where the number of observations per fold is relatively low. If you are using this function, it's highly recommended you update the package.

## glmnetUtils 1.1.1
- Fixes bug where `nfolds` argument was not being passed to `glmnet::cv.glmnet`.

## glmnetUtils 1.1
- Now allows interaction and expression terms without requiring `use.model.frame=TRUE`. This works in an additive fashion, ie the formula `~ a + b:c + d*e` is treated as consisting of three terms, `a`, `b:c` and `d*e` each of which is processed independently of the others. A dot in the formula includes all main effect terms, ie `~ . + a:b + f(x)` expands to `~ a + b + x + a:b + f(x)` (assuming a, b and x are the only columns in the data). Note that a formula like `~ (a + b) + (c + d)` will be treated as two terms, `a + b` and `c + d`.
- The call component of a `glmnet`/`cv.glmnet` object that uses the original matrix/vector interface is now useful.
- You can now explicitly specify the vector of crossvalidation folds (for the inner loop over lambda) when calling `cva.glmnet`.
- Correctly handle non-syntactic factor variables in a formula.

## glmnetUtils 1.0.2
- Initial release to CRAN.


tests/testthat.R
Sys.setenv(R_TESTS="")  # needed for cluster creation to work

library(testthat)
library(glmnetUtils)

test_check("glmnetUtils")


tests/testthat/test1_glmnet.R
context("glmnet")


# TODO: better separation of simpleMM, glmnet
test_that("method dispatch works", {
    x <- as.matrix(mtcars[-1])
    y <- mtcars$mpg
    mod00 <- glmnet::glmnet(x, y)
    mod0 <- glmnet(x, y)
    mod1 <- glmnet(mpg ~ ., data=mtcars)
    expect_s3_class(mod1, "glmnet.formula")
    expect_equal(mod00$beta, mod0$beta)
    expect_equal(mod0$beta, mod1$beta)
})


test_that("simple model matrix works", {
    mod1 <- glmnet(mpg ~ ., data=mtcars, use.model.frame=TRUE)
    mod2 <- glmnet(mpg ~ ., data=mtcars, use.model.frame=FALSE)
    expect_equal(mod1$beta, mod2$beta)
})


test_that("glmnet arguments work", {
    x <- as.matrix(iris[-5])
    y <- as.matrix(iris$Species)
    mod00 <- glmnet::glmnet(x, y, family="multinomial", alpha=0.5)
    mod0 <- glmnet(x, y, family="multinomial", alpha=0.5)
    mod1 <- glmnet(Species ~ ., data=iris, family="multinomial")
    expect_equivalent(mod00$beta, mod0$beta)
    expect_equivalent(mod0$beta, mod1$beta)

    data(Insurance, package="MASS")
    # don't worry about ordered factors for now
    class(Insurance$Group) <- "factor"
    class(Insurance$Age) <- "factor"
    x <- model.matrix(~ District + Group + Age, data=Insurance)[, -1]
    y <- Insurance$Claims
    ofs <- Insurance$Holders
    mod00 <- glmnet::glmnet(x, y, family="poisson", offset=log(ofs))
    mod0 <- glmnet(x, y, family="poisson", offset=log(ofs))
    mod1 <- glmnet(Claims ~ District + Group + Age, data=Insurance, family="poisson", offset=log(Holders),
                   use.model.frame=TRUE)
    expect_equivalent(mod00$beta, mod0$beta)
    expect_equivalent(mod0$beta, mod1$beta)
})


test_that("predict and coef work", {
    x <- as.matrix(iris[-5])
    y <- iris$Species
    mod1.0 <- glmnet(x, y, family="multinomial")
    mod1.1 <- glmnet(Species ~ ., data=iris, family="multinomial", use.model.frame=TRUE)
    mod1.2 <- glmnet(Species ~ ., data=iris, family="multinomial", use.model.frame=FALSE)
    pred1.0 <- predict(mod1.0, as.matrix(iris[-5]), type="class")
    pred1.1 <- predict(mod1.1, iris, type="class")
    pred1.2 <- predict(mod1.2, iris, type="class")
    expect_equivalent(pred1.0, pred1.1)  # expect_equivalent because dimnames not always preserved
    expect_equivalent(pred1.1, pred1.2)
    coef1.0 <- coef(mod1.0)
    coef1.1 <- coef(mod1.1)
    coef1.2 <- coef(mod1.2)
    expect_equal(coef1.0, coef1.1)
    expect_equal(coef1.1, coef1.2)

    x <- as.matrix(mtcars[-1])
    y <- mtcars$mpg
    mod2.0 <- glmnet(x, y, alpha=0.5)
    mod2.1 <- glmnet(mpg ~ ., data=mtcars, alpha=0.5, use.model.frame=TRUE)
    mod2.2 <- glmnet(mpg ~ ., data=mtcars, alpha=0.5, use.model.frame=FALSE)
    pred2.0 <- predict(mod2.0, x, s=1)
    pred2.1 <- predict(mod2.1, mtcars, s=1)
    pred2.2 <- predict(mod2.2, mtcars, s=1)
    expect_equivalent(pred2.0, pred2.1)
    expect_equivalent(pred2.1, pred2.2)
    coef2.0 <- coef(mod2.0, s=1)
    coef2.1 <- coef(mod2.1, s=1)
    coef2.2 <- coef(mod2.2, s=1)
    expect_equal(coef2.0, coef2.1)
    expect_equal(coef2.1, coef2.2)
})


test_that("prediction with NA works", {
    data(Boston, package="MASS")
    BostonNA <- Boston
    BostonNA[1, ] <- NA

    mod1.1 <- glmnet(medv ~ ., data=Boston, alpha=0, use.model.frame=TRUE)
    mod1.2 <- glmnet(medv ~ ., data=Boston, alpha=0, use.model.frame=FALSE)

    pred1.1 <- predict(mod1.1, BostonNA, na.action=na.exclude)
    pred1.2 <- predict(mod1.2, BostonNA, na.action=na.exclude)
    expect_equivalent(pred1.1, pred1.2)
    expect(all(!is.na(pred1.1)), "NAs found with na.exclude")

    pred1.3 <- predict(mod1.1, BostonNA, na.action=na.pass)
    pred1.4 <- predict(mod1.2, BostonNA, na.action=na.pass)
    expect_equivalent(pred1.3, pred1.4)
    expect(all(is.na(pred1.3[1, ])), "NAs dropped with na.pass")
})


test_that("matrix y works", {
    df <- data.frame(x1 = rnorm(100), x2 = rnorm(100), y = rbinom(100, 10, 0.5))
    df$ny <- 10 - df$y
    mod1.1 <- glmnet(cbind(ny, y) ~ x1 + x2, data = df, family = "binomial", use.model.frame = TRUE)
    mod1.2 <- glmnet(cbind(ny, y) ~ x1 + x2, data = df, family = "binomial", use.model.frame = FALSE)

    pred1.1 <- predict(mod1.1, df)
    pred1.2 <- predict(mod1.2, df)
    expect_equivalent(pred1.1, pred1.2)
})


tests/testthat/test7_family.R
context("Enhanced family arg")

# to avoid issues with 32-bit Solaris checks on CRAN
skip_on_os("solaris")


test_that("enhanced family arg works with glmnet", {
    x <- as.matrix(mtcars[-1])
    y <- mtcars$mpg
    mod0.1 <- glmnet::glmnet(x, y)
    mod0.2 <- glmnet::glmnet(x, y, family="gaussian")
    mod0.3 <- glmnet::glmnet(x, y, family=gaussian)
    mod0.4 <- glmnet::glmnet(x, y, family=gaussian())
    mod1.1 <- glmnet(mpg ~ ., data=mtcars)
    mod1.2 <- glmnet(mpg ~ ., data=mtcars, family="gaussian")
    mod1.3 <- glmnet(mpg ~ ., data=mtcars, family=gaussian)
    mod1.4 <- glmnet(mpg ~ ., data=mtcars, family=gaussian())
    expect_s3_class(mod1.1, "glmnet.formula")
    expect_s3_class(mod1.2, "glmnet.formula")
    expect_s3_class(mod1.3, "glmnet.formula")
    expect_s3_class(mod1.4, "glmnet.formula")
    expect_equal(mod0.1$beta, mod1.1$beta)
    expect_equal(mod0.2$beta, mod1.2$beta)
    expect_equal(mod0.3$beta, mod1.3$beta)
    expect_equal(mod0.4$beta, mod1.4$beta)
})


test_that("enhanced family arg works with cv.glmnet", {
    skip_if_not_installed("MASS")
    data(Boston, package="MASS")
    x <- as.matrix(Boston[-1])
    y <- Boston$crim

    set.seed(12345)
    mod0.1 <- glmnet::cv.glmnet(x, y, alpha=0.5)
    set.seed(12345)
    mod0.2 <- glmnet::cv.glmnet(x, y, alpha=0.5, family="gaussian")
    set.seed(12345)
    mod0.3 <- glmnet::cv.glmnet(x, y, alpha=0.5, family=gaussian)
    set.seed(12345)
    mod0.4 <- glmnet::cv.glmnet(x, y, alpha=0.5, family=gaussian())

    set.seed(12345)
    mod1.1 <- cv.glmnet(crim ~ ., data=Boston, alpha=0.5)
    set.seed(12345)
    mod1.2 <- cv.glmnet(crim ~ ., data=Boston, alpha=0.5, family="gaussian")
    set.seed(12345)
    mod1.3 <- cv.glmnet(crim ~ ., data=Boston, alpha=0.5, family=gaussian)
    set.seed(12345)
    mod1.4 <- cv.glmnet(crim ~ ., data=Boston, alpha=0.5, family=gaussian())

    expect_equal(mod0.1$beta, mod1.1$beta)
    expect_equal(mod0.2$beta, mod1.2$beta)
    expect_equal(mod0.3$beta, mod1.3$beta)
    expect_equal(mod0.4$beta, mod1.4$beta)
})


test_that("enhanced family arg works with cva.glmnet", {
    skip_if_not_installed("MASS")
    data(Boston, package="MASS")

    set.seed(12345)
    mod1.1 <- cva.glmnet(medv ~ ., data=Boston)
    set.seed(12345)
    mod1.2 <- cva.glmnet(medv ~ ., data=Boston, family="gaussian")
    set.seed(12345)
    mod1.3 <- cva.glmnet(medv ~ ., data=Boston, family=gaussian)
    set.seed(12345)
    mod1.4 <- cva.glmnet(medv ~ ., data=Boston, family=gaussian())

    w <- 5
    expect_equal(coef(mod1.1, which=w), coef(mod1.2, which=w))
    expect_equal(coef(mod1.3, which=w), coef(mod1.4, which=w))
})


test_that("enhanced family arg works with relaxed glmnet", {
    x <- as.matrix(mtcars[-1])
    y <- mtcars$mpg
    mod0.1 <- glmnet::glmnet(x, y, alpha=0.5, relax=TRUE)
    mod0.2 <- glmnet::glmnet(x, y, family="gaussian", alpha=0.5, relax=TRUE)
    mod0.3 <- glmnet::glmnet(x, y, family=gaussian, alpha=0.5, relax=TRUE)
    mod0.4 <- glmnet::glmnet(x, y, family=gaussian(), alpha=0.5, relax=TRUE)

    mod1.1 <- glmnet(mpg ~ ., data=mtcars, alpha=0.5, relax=TRUE)
    mod1.2 <- glmnet(mpg ~ ., data=mtcars, family="gaussian", alpha=0.5, relax=TRUE)
    mod1.3 <- glmnet(mpg ~ ., data=mtcars, family=gaussian, alpha=0.5, relax=TRUE)
    mod1.4 <- glmnet(mpg ~ ., data=mtcars, family=gaussian(), alpha=0.5, relax=TRUE)
    expect_equivalent(mod0.1$beta, mod1.1$beta)
    expect_equivalent(mod0.2$beta, mod1.2$beta)
    expect_equivalent(mod0.3$beta, mod1.3$beta)
    expect_equivalent(mod0.4$beta, mod1.4$beta)
})


test_that("enhanced family arg works with relaxed cv.glmnet", {
    skip_if_not_installed("MASS")
    data(Boston, package="MASS")
    x <- as.matrix(Boston[-1])
    y <- Boston$crim
    set.seed(12345)
    mod0.1 <- glmnet::cv.glmnet(x, y, alpha=0.5, gamma=(0:3)/3, relax=TRUE)
    set.seed(12345)
    mod0.2 <- glmnet::cv.glmnet(x, y, family="gaussian", alpha=0.5, gamma=(0:3)/3, relax=TRUE)
    set.seed(12345)
    mod0.3 <- glmnet::cv.glmnet(x, y, family=gaussian, alpha=0.5, gamma=(0:3)/3, relax=TRUE)
    set.seed(12345)
    mod0.4 <- glmnet::cv.glmnet(x, y, family=gaussian(), alpha=0.5, gamma=(0:3)/3, relax=TRUE)

    set.seed(12345)
    mod1.1 <- cv.glmnet(crim ~ ., data=Boston, alpha=0.5, gamma=(0:3)/3, relax=TRUE)
    set.seed(12345)
    mod1.2 <- cv.glmnet(crim ~ ., data=Boston, family="gaussian", alpha=0.5, gamma=(0:3)/3, relax=TRUE)
    set.seed(12345)
    mod1.3 <- cv.glmnet(crim ~ ., data=Boston, family=gaussian, alpha=0.5, gamma=(0:3)/3, relax=TRUE)
    set.seed(12345)
    mod1.4 <- cv.glmnet(crim ~ ., data=Boston, family=gaussian(), alpha=0.5, gamma=(0:3)/3, relax=TRUE)

    expect_equal(mod0.1$glmnet.fit$beta, mod1.1$glmnet.fit$beta)
    expect_equal(mod0.2$glmnet.fit$beta, mod1.2$glmnet.fit$beta)
    expect_equal(mod0.3$glmnet.fit$beta, mod1.3$glmnet.fit$beta)
    expect_equal(mod0.4$glmnet.fit$beta, mod1.4$glmnet.fit$beta)
})


R/glmnetUtils.r
#' @title Utilities for glmnet
#' @description
#' Some quality-of-life functions to streamline the process of fitting elastic net models with the `glmnet` package, specifically:
#'
#' * `glmnet.formula` provides a formula/data frame interface to `glmnet`.
#' * `cv.glmnet.formula` does a similar thing for `cv.glmnet`.
#' * Methods for `predict` and `coef` for both the above.
#' * A function `cva.glmnet` to choose both the alpha and lambda parameters via cross-validation, following the approach described in the help page for `cv.glmnet`. Optionally does the cross-validation in parallel.
#' * Methods for `plot`, `predict` and `coef` for the above.
#'
#' @docType package
#' @name glmnetUtils
#' @aliases glmnetUtils-package
NULL


## assorted imports from base packages
#' @importFrom graphics lines plot
#' @importFrom grDevices palette topo.colors
#' @importFrom stats coef na.pass predict terms .getXlevels delete.response
#' @importFrom stats model.frame model.extract model.matrix model.response
NULL


R/cvGlmnetFormula.r
#' @include glmnetUtils.r
NULL

#' @name cv.glmnet
#' @export
cv.glmnet <- function(x, ...)
UseMethod("cv.glmnet")

#' @rdname cv.glmnet
#' @method cv.glmnet default
#' @export
cv.glmnet.default <- function(x, y, ...)
{
    cl <- match.call()
    cl[[1]] <- quote(glmnet::cv.glmnet)
    obj <- glmnet::cv.glmnet(x, y, ...)
    obj$call <- cl
    obj
}


#' Formula interface for elastic net cross-validation with cv.glmnet
#'
#' @param x For the default method, a matrix of predictor variables.
#' @param y For the default method, a response vector or matrix (for a multinomial response).
#' @param formula A model formula; interaction terms are allowed and will be expanded per the usual rules for linear models.
#' @param data A data frame or matrix containing the variables in the formula.
#' @param weights An optional vector of case weights to be used in the fitting process. If missing, defaults to an unweighted fit.
#' @param offset An optional vector of offsets, an _a priori_ known component to be included in the linear predictor.
#' @param subset An optional vector specifying the subset of observations to be used to fit the model.
#' @param na.action A function which indicates what should happen when the data contains missing values. For the `predict` method, `na.action = na.pass` will predict missing values with `NA`; `na.omit` or `na.exclude` will drop them.
#' @param drop.unused.levels Should factors have unused levels dropped? Defaults to `FALSE`.
#' @param xlev A named list of character vectors giving the full set of levels to be assumed for each factor.
#' @param alpha The elastic net mixing parameter. See [glmnet::glmnet] for more details.
#' @param nfolds The number of crossvalidation folds to use. See [glmnet::cv.glmnet] for more details.
#' @param sparse Should the model matrix be in sparse format? This can save memory when dealing with many factor variables, each with many levels.
#' @param use.model.frame Should the base [model.frame] function be used when constructing the model matrix? This is the standard method that most R modelling functions use, but has some disadvantages. The default is to avoid `model.frame` and construct the model matrix term-by-term; see [discussion][glmnet.model.matrix].
#' @param gamma For `cv.glmnet.formula`, the values of the parameter for mixing the relaxed (non-regularised) fit with the regularized fit. Not used if `relax=FALSE`. Requires glmnet 3.0 or later.
#' @param relax For `cv.glmnet.formula`, whether to perform a relaxed fit after the regularised one. Requires glmnet 3.0 or later.
#' @param ... For `cv.glmnet.formula` and `cv.glmnet.default`, other arguments to be passed to [glmnet::cv.glmnet]; for the `predict` and `coef` methods, arguments to be passed to their counterparts in package glmnet.
#'
#' @details
#' The `cv.glmnet` function in this package is an S3 generic with a formula and a default method. The former calls the latter, and the latter is simply a direct call to the `cv.glmnet` function in package `glmnet`. All the arguments to `glmnet::cv.glmnet` are (or should be) supported.
#'
#' There are two ways in which the matrix of predictors can be generated. The default, with `use.model.frame = FALSE`, is to process the additive terms in the formula independently. With wide datasets, this is much faster and more memory-efficient than the standard R approach which uses the `model.frame` and `model.matrix` functions. However, the resulting model object is not exactly the same as if the standard approach had been used; in particular, it lacks a bona fide [terms] object. If you require interoperability with other packages that assume the standard model object structure, set `use.model.frame = TRUE`. See [discussion][glmnet.model.matrix] for more information on this topic.
#'
#' The `predict` and `coef` methods are wrappers for the corresponding methods in the glmnet package. The former constructs a predictor model matrix from its `newdata` argument and passes that as the `newx` argument to `glmnet:::predict.cv.glmnet`.
#'
#' @return
#' For `cv.glmnet.formula`, an object of class either `cv.glmnet.formula` or `cv.relaxed.formula`, based on the value of the `relax` argument. This is basically the same object created by `glmnet::cv.glmnet`, but with extra components to allow formula usage.
#'
#' @seealso
#' [glmnet::cv.glmnet], [glmnet::predict.cv.glmnet], [glmnet::coef.cv.glmnet], [model.frame], [model.matrix]
#'
#' @examples
#' cv.glmnet(mpg ~ ., data=mtcars)
#'
#' cv.glmnet(Species ~ ., data=iris, family="multinomial")
#'
#' \dontrun{
#'
#' # Leukemia example dataset from Trevor Hastie's website
#' download.file("https://web.stanford.edu/~hastie/glmnet/glmnetData/Leukemia.RData",
#'               "Leukemia.RData")
#' load("Leukemia.Rdata")
#' leuk <- do.call(data.frame, Leukemia)
#' cv.glmnet(y ~ ., leuk, family="binomial")
#' }
#' @rdname cv.glmnet
#' @method cv.glmnet formula
#' @importFrom glmnet cv.glmnet
#' @export
cv.glmnet.formula <- function(formula, data, alpha=1, nfolds=10, ..., weights=NULL, offset=NULL, subset=NULL,
                              na.action=getOption("na.action"), drop.unused.levels=FALSE, xlev=NULL,
                              sparse=FALSE, use.model.frame=FALSE, gamma=c(0, 0.25, 0.5, 0.75, 1), relax=FALSE)
{
    # must use NSE to get model.frame emulation to work
    cl <- match.call(expand.dots=FALSE)
    cl[[1]] <- if(use.model.frame)
        makeModelComponentsMF
    else makeModelComponents
    xy <- eval.parent(cl)

    if(relax && utils::packageVersion("glmnet") < package_version("3.0.0"))
        stop("Relaxed fit requires glmnet version 3.0 or higher", call.=FALSE)

    model <- glmnet::cv.glmnet(xy$x, xy$y, weights=xy$weights, offset=xy$offset, alpha=alpha,
                               nfolds=nfolds, ..., gamma=gamma, relax=relax)
    model$call <- match.call()
    model$terms <- xy$terms
    model$xlev <- xy$xlev
    model$alpha <- alpha
    model$nfolds <- nfolds
    model$sparse <- sparse
    model$use.model.frame <- use.model.frame
    model$na.action <- na.action
    class(model) <- if(relax)
        c("cv.relaxed.formula", class(model))
    else c("cv.glmnet.formula", class(model))
    model
}


#' @param object For the `predict` and `coef` methods, an object of class `cv.glmnet.formula`.
#' @param newdata For the `predict` method, a data frame containing the observations for which to calculate predictions.
#' @rdname cv.glmnet
#' @method predict cv.glmnet.formula
#' @export
predict.cv.glmnet.formula <- function(object, newdata, na.action=na.pass, ...)
{
    if(!inherits(object, "cv.glmnet.formula"))
        stop("invalid cv.glmnet.formula object")

    # must use NSE to get model.frame emulation to work
    cl <- match.call(expand.dots=FALSE)
    cl$formula <- delete.response(object$terms)
    cl$data <- cl$newdata
    cl$newdata <- NULL
    cl$xlev <- object$xlev
    cl[[1]] <- if(object$use.model.frame)
        makeModelComponentsMF
    else makeModelComponents

    xy <- eval.parent(cl)
    x <- xy$x
    offset <- xy$offset

    class(object) <- class(object)[-1]
    predict(object, x, ...)
}


#' @rdname cv.glmnet
#' @method coef cv.glmnet.formula
#' @export
coef.cv.glmnet.formula <- function(object, ...)
{
    if(!inherits(object, "cv.glmnet.formula"))
        stop("invalid cv.glmnet.formula object")
    class(object) <- class(object)[-1]
    coef(object, ...)
}


#' @rdname cv.glmnet
#' @method print cv.glmnet.formula
#' @export
print.cv.glmnet.formula <- function(x, ...)
{
    cat("Call:\n")
    dput(x$call)
    cat("\nModel fitting options:")
    cat("\n    Sparse model matrix:", x$sparse)
    cat("\n    Use model.frame:", x$use.model.frame)
    cat("\n    Number of crossvalidation folds:", x$nfolds)
    cat("\n    Alpha:", x$alpha)
    cat("\n    Deviance-minimizing lambda:", x$lambda.min, " (+1 SE):", x$lambda.1se)
    cat("\n")
    invisible(x)
}


#' @rdname cv.glmnet
#' @method predict cv.relaxed.formula
#' @export
predict.cv.relaxed.formula <- function(object, newdata, na.action=na.pass, ...)
{
    if(!inherits(object, "cv.relaxed.formula"))
        stop("invalid cv.relaxed.formula object")

    # must use NSE to get model.frame emulation to work
    cl <- match.call(expand.dots=FALSE)
    cl$formula <- delete.response(object$terms)
    cl$data <- cl$newdata
    cl$newdata <- NULL
    cl$xlev <- object$xlev
    cl[[1]] <- if(object$use.model.frame)
        makeModelComponentsMF
    else makeModelComponents

    xy <- eval.parent(cl)
    x <- xy$x
    offset <- xy$offset

    class(object) <- class(object)[-1]
    predict(object, x, ...)
}


#' @rdname cv.glmnet
#' @method coef cv.glmnet.formula
#' @export
coef.cv.relaxed.formula <- function(object, ...)
{
    if(!inherits(object, "cv.relaxed.formula"))
        stop("invalid cv.relaxed.formula object")
    class(object) <- class(object)[-1]
    coef(object, ...)
}


R/glmnetFormula.r
#' @include glmnetUtils.r
NULL

#' @name glmnet
#' @export
glmnet <- function(x, ...)
UseMethod("glmnet")

#' @rdname glmnet
#' @method glmnet default
#' @export
glmnet.default <- function(x, y, ...)
{
    cl <- match.call()
    cl[[1]] <- quote(glmnet::glmnet)
    obj <- glmnet::glmnet(x, y, ...)
    obj$call <- cl
    obj
}


#' Formula interface for elastic net modelling with glmnet
#'
#' @param x For the default method, a matrix of predictor variables.
#' @param y For the default method, a response vector or matrix (for a multinomial response).
#' @param formula A model formula; interaction terms are allowed and will be expanded per the usual rules for linear models.
#' @param data A data frame or matrix containing the variables in the formula.
#' @param family The model family. See [glmnet::glmnet] for how to specify this argument.
#' @param weights An optional vector of case weights to be used in the fitting process. If missing, defaults to an unweighted fit.
#' @param offset An optional vector of offsets, an _a priori_ known component to be included in the linear predictor.
#' @param subset An optional vector specifying the subset of observations to be used to fit the model.
#' @param na.action A function which indicates what should happen when the data contains missing values. For the `predict` method, `na.action = na.pass` will predict missing values with `NA`; `na.omit` or `na.exclude` will drop them.
#' @param drop.unused.levels Should factors have unused levels dropped? Defaults to `FALSE`.
#' @param xlev A named list of character vectors giving the full set of levels to be assumed for each factor.
#' @param alpha The elastic net mixing parameter. See [glmnet::glmnet] for more details.
#' @param sparse Should the model matrix be in sparse format? This can save memory when dealing with many factor variables, each with many levels.
#' @param use.model.frame Should the base [model.frame] function be used when constructing the model matrix? This is the standard method that most R modelling functions use, but has some disadvantages. The default is to avoid `model.frame` and construct the model matrix term-by-term; see [discussion][glmnet.model.matrix].
#' @param relax For `glmnet.formula`, whether to perform a relaxed (non-regularised) fit after the regularised one. Requires glmnet 3.0 or later.
#' @param ... For `glmnet.formula` and `glmnet.default`, other arguments to be passed to [glmnet::glmnet]; for the `predict` and `coef` methods, arguments to be passed to their counterparts in package glmnet.
#'
#' @details
#' The `glmnet` function in this package is an S3 generic with a formula and a default method. The former calls the latter, and the latter is simply a direct call to the `glmnet` function in package glmnet. All the arguments to `glmnet::glmnet` are (or should be) supported.
#'
#' There are two ways in which the matrix of predictors can be generated. The default, with `use.model.frame = FALSE`, is to process the additive terms in the formula independently. With wide datasets, this is much faster and more memory-efficient than the standard R approach which uses the `model.frame` and `model.matrix` functions. However, the resulting model object is not exactly the same as if the standard approach had been used; in particular, it lacks a bona fide [terms] object. If you require interoperability with other packages that assume the standard model object structure, set `use.model.frame = TRUE`. See [discussion][glmnet.model.matrix] for more information on this topic.
#'
#' The `predict` and `coef` methods are wrappers for the corresponding methods in the glmnet package. The former constructs a predictor model matrix from its `newdata` argument and passes that as the `newx` argument to `glmnet:::predict.glmnet`.
#'
#' @section Value:
#' For `glmnet.formula`, an object of class either `glmnet.formula` or `relaxed.formula`, based on the value of the `relax` argument. This is basically the same object created by `glmnet::glmnet`, but with extra components to allow formula usage.
#'
#' @seealso
#' [glmnet::glmnet], [glmnet::predict.glmnet], [glmnet::coef.glmnet], [model.frame], [model.matrix]
#'
#' @examples
#' glmnet(mpg ~ ., data=mtcars)
#'
#' glmnet(Species ~ ., data=iris, family="multinomial")
#'
#' \dontrun{
#'
#' # Leukemia example dataset from Trevor Hastie's website
#' download.file("https://web.stanford.edu/~hastie/glmnet/glmnetData/Leukemia.RData",
#'               "Leukemia.RData")
#' load("Leukemia.Rdata")
#' leuk <- do.call(data.frame, Leukemia)
#' glmnet(y ~ ., leuk, family="binomial")
#' }
#' @rdname glmnet
#' @method glmnet formula
#' @importFrom glmnet glmnet
#' @export
glmnet.formula <- function(formula, data,
                           family=c("gaussian", "binomial", "poisson", "multinomial", "cox", "mgaussian"),
                           alpha=1, ..., weights=NULL, offset=NULL, subset=NULL,
                           na.action=getOption("na.action"), drop.unused.levels=FALSE, xlev=NULL,
                           sparse=FALSE, use.model.frame=FALSE, relax=FALSE)
{
    # must use NSE to get model.frame emulation to work
    cl <- match.call(expand.dots=FALSE)
    cl[[1]] <- if(use.model.frame)
        makeModelComponentsMF
    else makeModelComponents
    xy <- eval.parent(cl)

    if(is.character(family))
        family <- match.arg(family)
    else
    {
        if(utils::packageVersion("glmnet") < package_version("4.0.0"))
            stop("Enhanced family argument requires glmnet version 4.0 or higher", call.=FALSE)
        if(is.function(family))
            family <- family()
        else if(!inherits(family, "family"))
            stop("Invalid family argument; must be either character, function or family object")
    }

    model <- glmnet::glmnet(xy$x, y=xy$y, family=family, weights=xy$weights, offset=xy$offset, alpha=alpha, ...)
    model$call <- match.call()
    model$call[[1]] <- parse(text="glmnetUtils:::glmnet.formula")[[1]]  # needed to make relaxed fitting work
    model$terms <- xy$terms
    model$xlev <- xy$xlev
    model$alpha <- alpha
    model$sparse <- sparse
    model$use.model.frame <- use.model.frame
    model$na.action <- na.action

    # can't pass relax directly to glmnet because of NSE wackiness induced by update()
    if(relax)
    {
        if(utils::packageVersion("glmnet") < package_version("3.0.0"))
            stop("Relaxed fit requires glmnet version 3.0 or higher", call.=FALSE)
        model <- glmnet::relax.glmnet(model, xy$x, family=family, weights=xy$weights, offset=xy$offset, alpha=alpha,
                                      ..., check.args=FALSE)
        class(model) <- c("relaxed.formula", class(model))
    }
    else class(model) <- c("glmnet.formula", class(model))

    model
}


#' @param object For the `predict` and `coef` methods, an object of class `glmnet.formula`.
#' @param newdata For the `predict` method, a data frame containing the observations for which to calculate predictions.
#' @rdname glmnet
#' @importFrom Matrix sparse.model.matrix
#' @export
#' @method predict glmnet.formula
predict.glmnet.formula <- function(object, newdata, offset=NULL, na.action=na.pass, ...)
{
    if(!inherits(object, "glmnet.formula"))
        stop("invalid glmnet.formula object")

    # must use NSE to get model.frame emulation to work
    cl <- match.call(expand.dots=FALSE)
    cl$formula <- delete.response(object$terms)
    cl$data <- cl$newdata
    cl$newdata <- NULL
    cl$xlev <- object$xlev
    cl[[1]] <- if(object$use.model.frame)
        makeModelComponentsMF
    else makeModelComponents

    xy <- eval.parent(cl)
    x <- xy$x
    offset <- xy$offset

    class(object) <- class(object)[-1]
    predict(object, x, offset=offset, ...)
}

#' @rdname glmnet
#' @export
#' @method coef glmnet.formula
coef.glmnet.formula <- function(object, ...)
{
    if(!inherits(object, "glmnet.formula"))
        stop("invalid glmnet.formula object")
    class(object) <- class(object)[-1]
    coef(object, ...)
}


#' @param digits Significant digits in printed output.
#' @param print.deviance.ratios Whether to print the table of deviance ratios, as per [glmnet::print.glmnet].
#' @rdname glmnet
#' @export
#' @method print glmnet.formula
print.glmnet.formula <- function(x, digits=max(3, getOption("digits") - 3), print.deviance.ratios=FALSE, ...)
{
    cat("Call:\n")
    dput(x$call)
    cat("\nModel fitting options:")
    cat("\n    Sparse model matrix:", x$sparse)
    cat("\n    Use model.frame:", x$use.model.frame)
    cat("\n    Alpha:", x$alpha)
    cat("\n    Lambda summary:\n")
    print(summary(x$lambda))
    if(print.deviance.ratios)
    {
        cat("\nDeviance ratios:\n")
        print(cbind(Df=x$df, `%Dev`=signif(x$dev.ratio, digits), Lambda=signif(x$lambda, digits)))
    }
    cat("\n")
    invisible(x)
}


#' @rdname glmnet
#' @export
#' @method print relaxed.formula
print.relaxed.formula <- function(x, digits=max(3, getOption("digits") - 3), print.deviance.ratios=FALSE, ...)
{
    print.glmnet.formula(x)
    cat("Relaxed fit in component $relaxed\n")
    invisible(x)
}


#' @rdname glmnet
#' @export
#' @method predict relaxed.formula
predict.relaxed.formula <- function(object, newdata, offset=NULL, na.action=na.pass, ...)
{
    if(!inherits(object, "relaxed.formula"))
        stop("invalid relaxed.formula object")

    # must use NSE to get model.frame emulation to work
    cl <- match.call(expand.dots=FALSE)
    cl$formula <- delete.response(object$terms)
    cl$data <- cl$newdata
    cl$newdata <- NULL
    cl$xlev <- object$xlev
    cl[[1]] <- if(object$use.model.frame)
        makeModelComponentsMF
    else makeModelComponents

    xy <- eval.parent(cl)
    x <- xy$x
    offset <- xy$offset

    class(object) <- class(object)[-1]
    predict(object, x, offset=offset, ...)
}

#' @rdname glmnet
#' @export
#' @method coef relaxed.formula
coef.relaxed.formula <- function(object, ...)
{
    if(!inherits(object, "relaxed.formula"))
        stop("invalid relaxed.formula object")
    class(object) <- class(object)[-1]
    coef(object, ...)
}
